File: frontend/src/api/aiPredictionDashboard.ts
import axios from 'axios';

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000';

const api = axios.create({
  baseURL: `${API_BASE_URL}/api/v1`,
  headers: {
    'Content-Type': 'application/json',
  },
});

api.interceptors.request.use((config) => {
  const token = localStorage.getItem('token');
  if (token) {
    config.headers.Authorization = `Bearer ${token}`;
  }
  return config;
});

export interface TopicProbabilityRanking {
  topic_id?: number;
  topic_name: string;
  chapter_name?: string;
  probability_score: number;
  star_rating: number;
  confidence_level: string;
  frequency_count: number;
  last_appeared_year?: number;
  years_since_last_appearance: number;
  is_due: boolean;
  priority_tag: string;
  expected_marks: number;
  study_hours_recommended: number;
}

export interface QuestionPaperSection {
  section_name: string;
  total_marks: number;
  question_types: string[];
  topics_included: string[];
  difficulty_distribution: Record<string, number>;
  bloom_level_distribution: Record<string, number>;
}

export interface PredictedQuestionBlueprint {
  total_marks: number;
  duration_minutes: number;
  sections: QuestionPaperSection[];
  topic_coverage: Record<string, number>;
  difficulty_breakdown: Record<string, number>;
}

export interface MarksDistribution {
  category: string;
  marks: number;
  percentage: number;
  color: string;
}

export interface FocusAreaRecommendation {
  topic_id?: number;
  topic_name: string;
  chapter_name?: string;
  priority: string;
  priority_score: number;
  reason: string;
  expected_impact: string;
  study_hours_needed: number;
  resources: string[];
  difficulty_level: string;
}

export interface StudyTimeAllocation {
  category: string;
  hours: number;
  percentage: number;
  color: string;
  description: string;
}

export interface AIPredictionDashboardResponse {
  board: string;
  grade_id: number;
  subject_id: number;
  subject_name: string;
  generated_at: string;
  topic_rankings: TopicProbabilityRanking[];
  predicted_blueprint: PredictedQuestionBlueprint;
  marks_distribution: MarksDistribution[];
  focus_areas: FocusAreaRecommendation[];
  study_time_allocation: StudyTimeAllocation[];
  overall_prediction: Record<string, number | string>;
}

export interface DailyTask {
  task_id: string;
  date: string;
  topic_name: string;
  task_type: string;
  duration_hours: number;
  priority: string;
  description: string;
  resources: string[];
  is_completed: boolean;
}

export interface StudyPlanWeek {
  week_number: number;
  start_date: string;
  end_date: string;
  focus_topics: string[];
  total_hours: number;
  tasks: DailyTask[];
}

export interface StudyPlanRequest {
  board: string;
  grade_id: number;
  subject_id: number;
  exam_date: string;
  available_hours_per_day: number;
  weak_areas?: number[];
}

export interface StudyPlanResponse {
  exam_date: string;
  days_until_exam: number;
  total_study_hours: number;
  weeks: StudyPlanWeek[];
  completion_percentage: number;
  milestone_dates: Record<string, string>;
}

export interface WhatIfScenarioRequest {
  board: string;
  grade_id: number;
  subject_id: number;
  study_hours_adjustment: number;
  focus_topic_ids?: number[];
  practice_test_count: number;
}

export interface PredictionChange {
  metric: string;
  current_value: number;
  projected_value: number;
  change_percentage: number;
  impact_level: string;
}

export interface WhatIfScenarioResponse {
  current_predicted_score: number;
  projected_score: number;
  score_improvement: number;
  confidence_level: string;
  prediction_changes: PredictionChange[];
  recommended_adjustments: string[];
  risk_factors: string[];
}

export interface CrashCourseTopicPriority {
  topic_id?: number;
  topic_name: string;
  priority_level: number;
  time_to_study_hours: number;
  expected_marks: number;
  roi_score: number;
  quick_revision_points: string[];
  must_know_concepts: string[];
  practice_questions: string[];
}

export interface CrashCourseDay {
  day_number: number;
  date: string;
  morning_session: string[];
  afternoon_session: string[];
  evening_session: string[];
  revision_topics: string[];
  practice_tests: string[];
  total_hours: number;
}

export interface CrashCourseModeResponse {
  days_until_exam: number;
  mode_activated: boolean;
  priority_topics: CrashCourseTopicPriority[];
  daily_schedule: CrashCourseDay[];
  quick_wins: string[];
  topics_to_skip: string[];
  estimated_coverage: number;
  expected_score_range: Record<string, number>;
}

const aiPredictionDashboardApi = {
  getDashboard: async (
    board: string,
    gradeId: number,
    subjectId: number
  ): Promise<AIPredictionDashboardResponse> => {
    const response = await api.get('/ai-prediction-dashboard/dashboard', {
      params: { board, grade_id: gradeId, subject_id: subjectId },
    });
    return response.data;
  },

  generateStudyPlan: async (request: StudyPlanRequest): Promise<StudyPlanResponse> => {
    const response = await api.post('/ai-prediction-dashboard/study-plan', request);
    return response.data;
  },

  simulateWhatIfScenario: async (
    request: WhatIfScenarioRequest
  ): Promise<WhatIfScenarioResponse> => {
    const response = await api.post('/ai-prediction-dashboard/what-if-scenario', request);
    return response.data;
  },

  activateCrashCourseMode: async (
    board: string,
    gradeId: number,
    subjectId: number,
    daysUntilExam: number
  ): Promise<CrashCourseModeResponse> => {
    const response = await api.get('/ai-prediction-dashboard/crash-course-mode', {
      params: {
        board,
        grade_id: gradeId,
        subject_id: subjectId,
        days_until_exam: daysUntilExam,
      },
    });
    return response.data;
  },
};

export default aiPredictionDashboardApi;

File: src/schemas/ai_prediction_dashboard.py
from datetime import datetime, date
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field
from src.models.previous_year_papers import Board


class TopicProbabilityRanking(BaseModel):
    """Topic probability ranking with star ratings and percentage bars"""
    topic_id: Optional[int] = None
    topic_name: str
    chapter_name: Optional[str] = None
    probability_score: float = Field(..., ge=0, le=100)
    star_rating: int = Field(..., ge=1, le=5)
    confidence_level: str
    frequency_count: int
    last_appeared_year: Optional[int] = None
    years_since_last_appearance: int
    is_due: bool
    priority_tag: str
    expected_marks: float
    study_hours_recommended: float


class QuestionPaperSection(BaseModel):
    """Section in predicted question paper blueprint"""
    section_name: str
    total_marks: int
    question_types: List[str]
    topics_included: List[str]
    difficulty_distribution: Dict[str, int]
    bloom_level_distribution: Dict[str, int]


class PredictedQuestionBlueprint(BaseModel):
    """Predicted question paper blueprint with expandable sections"""
    total_marks: int
    duration_minutes: int
    sections: List[QuestionPaperSection]
    topic_coverage: Dict[str, float]
    difficulty_breakdown: Dict[str, int]


class MarksDistribution(BaseModel):
    """Expected marks distribution for pie chart"""
    category: str
    marks: float
    percentage: float
    color: str


class FocusAreaRecommendation(BaseModel):
    """Focus area recommendation with priority tags"""
    topic_id: Optional[int] = None
    topic_name: str
    chapter_name: Optional[str] = None
    priority: str  # "critical", "high", "medium", "low"
    priority_score: float
    reason: str
    expected_impact: str
    study_hours_needed: float
    resources: List[str]
    difficulty_level: str


class StudyTimeAllocation(BaseModel):
    """Study time allocation for donut chart"""
    category: str
    hours: float
    percentage: float
    color: str
    description: str


class DailyTask(BaseModel):
    """Daily task in personalized study plan"""
    task_id: str
    date: date
    topic_name: str
    task_type: str  # "study", "practice", "revision", "test"
    duration_hours: float
    priority: str
    description: str
    resources: List[str]
    is_completed: bool = False


class StudyPlanWeek(BaseModel):
    """Week in study plan timeline"""
    week_number: int
    start_date: date
    end_date: date
    focus_topics: List[str]
    total_hours: float
    tasks: List[DailyTask]


class StudyPlanRequest(BaseModel):
    """Request to generate personalized study plan"""
    board: Board
    grade_id: int
    subject_id: int
    exam_date: date
    available_hours_per_day: float = Field(..., ge=0.5, le=12)
    weak_areas: Optional[List[int]] = Field(None, description="List of topic IDs student struggles with")


class StudyPlanResponse(BaseModel):
    """Personalized study plan timeline with daily tasks"""
    exam_date: date
    days_until_exam: int
    total_study_hours: float
    weeks: List[StudyPlanWeek]
    completion_percentage: float
    milestone_dates: Dict[str, date]


class WhatIfScenarioRequest(BaseModel):
    """What-if scenario simulation request"""
    board: Board
    grade_id: int
    subject_id: int
    study_hours_adjustment: float = Field(0, description="Additional study hours per day (-5 to +10)")
    focus_topic_ids: Optional[List[int]] = Field(None, description="Topics to focus on")
    practice_test_count: int = Field(0, ge=0, le=20, description="Number of practice tests")


class PredictionChange(BaseModel):
    """Change in prediction based on what-if scenario"""
    metric: str
    current_value: float
    projected_value: float
    change_percentage: float
    impact_level: str


class WhatIfScenarioResponse(BaseModel):
    """What-if scenario simulation results"""
    current_predicted_score: float
    projected_score: float
    score_improvement: float
    confidence_level: str
    prediction_changes: List[PredictionChange]
    recommended_adjustments: List[str]
    risk_factors: List[str]


class CrashCourseTopicPriority(BaseModel):
    """Topic priority for crash course mode"""
    topic_id: Optional[int] = None
    topic_name: str
    priority_level: int  # 1-5, with 1 being highest
    time_to_study_hours: float
    expected_marks: float
    roi_score: float  # Return on investment score
    quick_revision_points: List[str]
    must_know_concepts: List[str]
    practice_questions: List[str]


class CrashCourseDay(BaseModel):
    """Daily plan in crash course mode"""
    day_number: int
    date: date
    morning_session: List[str]
    afternoon_session: List[str]
    evening_session: List[str]
    revision_topics: List[str]
    practice_tests: List[str]
    total_hours: float


class CrashCourseModeResponse(BaseModel):
    """Last-minute crash course mode activation response"""
    days_until_exam: int
    mode_activated: bool
    priority_topics: List[CrashCourseTopicPriority]
    daily_schedule: List[CrashCourseDay]
    quick_wins: List[str]
    topics_to_skip: List[str]
    estimated_coverage: float
    expected_score_range: Dict[str, float]


class AIPredictionDashboardResponse(BaseModel):
    """Complete AI prediction dashboard response"""
    board: Board
    grade_id: int
    subject_id: int
    subject_name: str
    generated_at: datetime
    
    # Topic probability rankings
    topic_rankings: List[TopicProbabilityRanking]
    
    # Predicted question paper blueprint
    predicted_blueprint: PredictedQuestionBlueprint
    
    # Expected marks distribution
    marks_distribution: List[MarksDistribution]
    
    # Focus area recommendations
    focus_areas: List[FocusAreaRecommendation]
    
    # Study time allocation
    study_time_allocation: List[StudyTimeAllocation]
    
    # Overall predictions
    overall_prediction: Dict[str, Any]
